### PR TITLE
Add diagnostic module

### PR DESCRIPTION
**Pull request by rzezeski**
Add a diagnostic module which gathers important information from
various parts of Yokozuna. This module will act as a first step for
sanity checking when issues occur in production. Currently it only
returns raw Erlang terms but other tools can use these raw terms to
build terminal output, GUIs, or hook into tools like Riaknostic.

This is the last step of #146.

### CODE REVIEW

**Comment by rzezeski** [comment]
This can be tested by running one of the riak tests, attaching to the console, and running `yz_diag:all().`.

Example output after running yz_pb test:

```
(dev1@127.0.0.1)2> rp(yz_diag:all()).
[{error_doc_counts,[{<<"basic">>,0},
                    {<<"stored_fields">>,0},
                    {<<"encoded">>,0}]},
 {indexes_in_meta,[<<"basic">>,<<"stored_fields">>,
                   <<"encoded">>]},
 {service_avail,['dev4@127.0.0.1','dev3@127.0.0.1',
                 'dev2@127.0.0.1','dev1@127.0.0.1']},
 {'dev4@127.0.0.1',[{component_status,[{index,true},
                                       {search,true}]},
                    {indexes_in_solr,[<<"basic">>,<<"encoded">>,
                                      <<"stored_fields">>]},
                    {searcher_stats,[{<<"basic">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,7},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:31.009Z">>},
                                       {warmup_time,0}]},
                                     {<<"encoded">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,1},
                                       {max_doc,0},
                                       {num_docs,0},
                                       {opened,<<"2014-04-03T20:33:30.991Z">>},
                                       {warmup_time,0}]},
                                     {<<"stored_fields">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:31.008Z">>},
                                       {warmup_time,0}]}]}]},
 {'dev3@127.0.0.1',[{component_status,[{index,true},
                                       {search,true}]},
                    {indexes_in_solr,[<<"basic">>,<<"encoded">>,
                                      <<"stored_fields">>]},
                    {searcher_stats,[{<<"basic">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:27.565Z">>},
                                       {warmup_time,0}]},
                                     {<<"encoded">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:27.565Z">>},
                                       {warmup_time,0}]},
                                     {<<"stored_fields">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:27.565Z">>},
                                       {warmup_time,0}]}]}]},
 {'dev2@127.0.0.1',[{component_status,[{index,true},
                                       {search,true}]},
                    {indexes_in_solr,[<<"basic">>,<<"encoded">>,
                                      <<"stored_fields">>]},
                    {searcher_stats,[{<<"basic">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:24.022Z">>},
                                       {warmup_time,0}]},
                                     {<<"encoded">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:24.022Z">>},
                                       {warmup_time,0}]},
                                     {<<"stored_fields">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,1},
                                       {max_doc,0},
                                       {num_docs,0},
                                       {opened,<<"2014-04-03T20:33:24.003Z">>},
                                       {warmup_time,0}]}]}]},
 {'dev1@127.0.0.1',[{component_status,[{index,true},
                                       {search,true}]},
                    {indexes_in_solr,[<<"basic">>,<<"encoded">>,
                                      <<"stored_fields">>]},
                    {searcher_stats,[{<<"basic">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,5},
                                       {max_doc,2},
                                       {num_docs,2},
                                       {opened,<<"2014-04-03T20:33:20.427Z">>},
                                       {warmup_time,0}]},
                                     {<<"encoded">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:20.425Z">>},
                                       {warmup_time,0}]},
                                     {<<"stored_fields">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,3},
                                       {max_doc,1},
                                       {num_docs,1},
                                       {opened,<<"2014-04-03T20:33:20.424Z">>},
                                       {warmup_time,0}]}]}]}]
ok
```

File: src/yz_diag.erl [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+%%
+%% @doc This module provides visibility into the state of Yokozuna. It
+%% is meant to aid in diagnosing Yokozuna when problems arise.
+%%
+%% All API functions should return Erlang data structures. They SHOULD
+%% NOT make logging calls or write to stdout/stderr. There may be
+%% functions which do that but they should be separate; built on top
+%% of these functions. It is very important to return Erlang data
+%% structures which are properly documented and spec'd so that they
+%% can be used by different consumers.
```
Comment (reiddraper) [inline src/yz_diag.erl:29]: nice doc!

File: src/yz_diag.erl [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+-module(yz_diag).
+-compile(export_all).
+-include("yokozuna.hrl").
+-type either(Term) :: Term | {error, Reason :: term()}.
+-type either_t(Term) :: Term | {error, Reason :: term(), StackTrace :: term()}.
```
Comment (reiddraper) [inline src/yz_diag.erl:34]: does the `_t` stand for trace?

File: src/yz_diag.erl [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                {error, _} = Err ->
+                    {Index, Err};
+                {_Headers, Resp} ->
+                    Struct = mochijson2:decode(Resp),
+                    Count = kvc:path([<<"response">>, <<"numFound">>], Struct),
+                    {Index, Count}
+            end
+    catch
+        _:Reason ->
+            {Index, Reason}
```
Comment (reiddraper) [inline src/yz_diag.erl:194]: this is a case we don't care about the stack trace, yeah?

**Comment by reiddraper** [comment]
Just a few minor questions but otherwise looking good. I'm going to go ahead and +1 the sha to kick off the integration build.

**Comment by rzezeski** [comment]
Develop was broken at some point due to upstream security changes, I believe. Once #363 is merged I'll rebase this.

File: src/yz_diag.erl (hunk 2 above)
Comment (rzezeski) [inline src/yz_diag.erl:34]: yep

File: src/yz_diag.erl (hunk 3 above)
Comment (rzezeski) [inline src/yz_diag.erl:194]: Yea, I explicitly decided to ignore it in this case to prevent inflating the returned structure.

**Comment by reiddraper** [comment]
> Once #363 is merged I'll rebase this.

Sounds good. I'll +1 the rebased sha once its up here.

**Comment by rzezeski** [comment]
I merged develop and tested that it still works:

```
(dev1@127.0.0.1)1> rp(yz_diag:all()).
[{error_doc_counts,[{<<"myindex">>,0},
                    {<<"myindex2">>,0},
                    {<<"h_myindex">>,0}]},
 {indexes_in_meta,[<<"myindex">>,<<"myindex2">>,
                   <<"h_myindex">>]},
 {service_avail,['dev1@127.0.0.1']},
 {'dev1@127.0.0.1',[{component_status,[{index,true},
                                       {search,true}]},
                    {indexes_in_solr,[<<"h_myindex">>,<<"myindex">>,
                                      <<"myindex2">>]},
                    {searcher_stats,[{<<"h_myindex">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,1},
                                       {max_doc,0},
                                       {num_docs,0},
                                       {opened,<<"2014-04-22T14:53:23.591Z">>},
                                       {warmup_time,0}]},
                                     {<<"myindex">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,1},
                                       {max_doc,0},
                                       {num_docs,0},
                                       {opened,<<"2014-04-22T14:53:16.048Z">>},
                                       {warmup_time,0}]},
                                     {<<"myindex2">>,
                                      [{caching,true},
                                       {deleted_docs,0},
                                       {index_version,1},
                                       {max_doc,0},
                                       {num_docs,0},
                                       {opened,<<"2014-04-22T14:53:16.048Z">>},
                                       {warmup_time,0}]}]}]}]
ok
```

**Comment by rzezeski** [comment]
@borshop merge

**Pull request merged**